### PR TITLE
Run CI on experimental `*.md` doc files

### PR DESCRIPTION
**Pull request by joschmitt**
The CI doctests would not run on the examples included in any `*.md` file inside `experimental` as discovered in https://oscar-system.slack.com/archives/C5B2H11BQ/p1726048480750199 .
This tries to fix this by linking the experimental documentation into `docs/src` as part of the CI runners. Currently, the tests are expected to fail because of missing `DocTestSetup = Oscar.doctestsetup()` lines in some of the previously uncovered files. If this works and people are happy with my changes, I will fix those doctests.

### CODE REVIEW

**Comment by benlorenz** [comment]
Could we move the `link_experimental_docs` function to `src/utils/docs.jl` and call it from `doctestsetup` instead, something like this:
```julia
function doctestsetup()
  link_experimental_docs(Oscar)
  return :(using Oscar; Oscar.AbstractAlgebra.set_current_module(@__MODULE__))
end
```
This would simplify running this locally and from OscarDevTools.

**Comment by thofma** [comment]
I am puzzled about the missing setup. Shouldn't `DocMeta.setdocmeta!(Oscar, :DocTestSetup, Oscar.doctestsetup(); recursive = true)` take care of that?

Edit: It probably only takes care of the doctests attached to functions, but not to the doctests in the `*md` files?

**Comment by joschmitt** [comment]
> I am puzzled about the missing setup. Shouldn't `DocMeta.setdocmeta!(Oscar, :DocTestSetup, Oscar.doctestsetup(); recursive = true)` take of that?

Yeah, I always wondered why we have to put this in every file.

**Comment by benlorenz** [comment]
`setdocmeta!` only applies to docstrings attached to functions and not to the markdown files. But I don't really know why it was designed like that.
There is an open issue for a better solution for the markdown files: https://github.com/JuliaDocs/Documenter.jl/issues/2512

**Comment by joschmitt** [comment]
> Could we move the `link_experimental_docs` function to `src/utils/docs.jl` and call it from `doctestsetup` instead, something like this:
> 
> ```julia
> function doctestsetup()
>   link_experimental_docs(Oscar)
>   return :(using Oscar; Oscar.AbstractAlgebra.set_current_module(@__MODULE__))
> end
> ```
> 
> This would simplify running this locally and from OscarDevTools.

I tried this now. If I put the CI runners back to the original state, they don't test the `*.md` files if `link_experimental_docs` is only done with `DocMeta.setdocmeta!(Oscar, :DocTestSetup, Oscar.doctestsetup(); recursive = true)`. The links are created, but I suppose they are created too late.

**Comment by benlorenz** [comment]
> > Could we move the `link_experimental_docs` function to `src/utils/docs.jl` and call it from `doctestsetup` instead, something like this:
> > ```julia
> > function doctestsetup()
> >   link_experimental_docs(Oscar)
> >   return :(using Oscar; Oscar.AbstractAlgebra.set_current_module(@__MODULE__))
> > end
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This would simplify running this locally and from OscarDevTools.
> 
> I tried this now. If I put the CI runners back to the original state, they don't test the `*.md` files if `link_experimental_docs` is only done with `DocMeta.setdocmeta!(Oscar, :DocTestSetup, Oscar.doctestsetup(); recursive = true)`. The links are created, but I suppose they are created too late.

Did you maybe put the call inside the quoted expression? Otherwise this doesn't seem to make sense to me, but I will give it a try myself.

**Comment by joschmitt** [comment]
> > > Could we move the `link_experimental_docs` function to `src/utils/docs.jl` and call it from `doctestsetup` instead, something like this:
> > > ```julia
> > > function doctestsetup()
> > >   link_experimental_docs(Oscar)
> > >   return :(using Oscar; Oscar.AbstractAlgebra.set_current_module(@__MODULE__))
> > > end
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > This would simplify running this locally and from OscarDevTools.
> > 
> > 
> > I tried this now. If I put the CI runners back to the original state, they don't test the `*.md` files if `link_experimental_docs` is only done with `DocMeta.setdocmeta!(Oscar, :DocTestSetup, Oscar.doctestsetup(); recursive = true)`. The links are created, but I suppose they are created too late.
> 
> Did you maybe put the call inside the quoted expression? Otherwise this doesn't seem to make sense to me, but I will give it a try myself.

I pushed now what I have.

**Comment by benlorenz** [comment]
Thanks, I understand it now: Many markdown files have `DocTestSetup = Oscar.doctestsetup()` and the linking step will first clean up all links and recreate them so this might cause issues / race-conditions (In general it seems reasonable to clean up old / bad links first).

The issue with having extra code in the yml files is that it would need more extra treatment for Oscar to get this to run in the downstream tests in OscarDevTools.

**Comment by joschmitt** [comment]
I try to make the "cleaning up" step of `link_experimental_docs` a bit less aggressive.

File: src/utils/docs.jl [hunk 1]
```diff
@@ -231,3 +234,33 @@ function build_doc(; doctest::Union{Symbol, Bool} = false, warnonly = true, open
     @warn versionwarn
   end
 end
+
+# Create symbolic links from any documentation directory in `experimental` into
+# `docs/src`
+function link_experimental_docs(Oscar::Module)
```
Comment (benlorenz) [inline src/utils/docs.jl]:
```suggestion
const _linked_exp_docs = Ref{Bool}(false)

function link_experimental_docs(Oscar::Module)
  Oscar._linked_exp_docs[] && return
```

File: src/utils/docs.jl [hunk 2]
```diff
@@ -231,3 +234,33 @@ function build_doc(; doctest::Union{Symbol, Bool} = false, warnonly = true, open
...
+    if !ispath(symlink_link)
+      symlink(symlink_target, symlink_link)
+    elseif !islink(symlink_link) || readlink(symlink_link) != symlink_target
+      error("""$symlink_link already exists, but is not a symlink to $symlink_target
+      Please investigate the contents of $symlink_link,
+      optionally move them somewhere else and delete the directory once you are done.""")
+    end
+  end
+
+  return nothing
```
Comment (benlorenz) [inline src/utils/docs.jl:270]:
```suggestion
  Oscar._linked_exp_docs[] = true
  return nothing
```

**Review by benlorenz — COMMENTED** [review]
We could also add a flag to make sure the linking is only done once.

**Comment by joschmitt** [comment]
Somehow nothing seems to work. I just pushed something to see the CI run. Please ignore it.

**Comment by joschmitt** [comment]
So the reason why the first attempt worked, but nothing else is this magical line in `docs/make_work.jl`:
```
# We monkey-patch Base.walkdir to use true as default value for follow_symlinks
# (normally false is the default), in order to "trick" the Documenter code into
# following those symlinks.
# See also:
# https://github.com/JuliaDocs/Documenter.jl/pull/552
# https://github.com/JuliaLang/julia/blob/master/doc/make.jl#L19
Base.walkdir(str::String) = Base.walkdir(str; follow_symlinks=true)
```
That's why it works if one includes `docs/make_work.jl`... :man_facepalming:

File: docs/make_work.jl [hunk 3]
```diff
@@ -26,14 +26,6 @@ include("citation_style.jl")
...
 Base.print(io::IO, b::Base.Docs.Binding) = print(io, b.var)
 
-# We monkey-patch Base.walkdir to use true as default value for follow_symlinks
-# (normally false is the default), in order to "trick" the Documenter code into
-# following those symlinks.
-# See also:
-# https://github.com/JuliaDocs/Documenter.jl/pull/552
-# https://github.com/JuliaLang/julia/blob/master/doc/make.jl#L19
-Base.walkdir(str::String) = Base.walkdir(str; follow_symlinks=true)
-
```
Comment (joschmitt) [inline docs/make_work.jl:36]: Note that moving this to `docs/documenter_helpers.jl` should be fine because that file is included in `docs/make_work.jl`.

**Comment by joschmitt** [comment]
> We could also add a flag to make sure the linking is only done once.

I think we introduced the deleting of all symlinks in https://github.com/oscar-system/Oscar.jl/pull/2215 because I started to switch branches and then some links weren't there anymore. With a global flag we would basically do the linking once per OSCAR session; I'm not sure that is stable in interactive use with Revise etc.?

For now, I edited `link_experimental_docs` to only delete links which shouldn't be there (and make sure that existing links are correct).

**Comment by joschmitt** [comment]
I have a version now which does not need to change the CI configuration. Could you have a look again @benlorenz?

**Comment by benlorenz** [comment]
Thanks for working on this! I think the test-code is good now.

Due to the `walkdir` issue this unfortunately doesn't work directly in OscarDevTools but should be simple enough to add by conditionally including `documenter_helpers.jl` (which I did not consider until now since it was only cosmetic).

**Review by benlorenz — APPROVED** [review]

File: docs/documenter_helpers.jl [hunk 4]
```diff
@@ -23,3 +23,11 @@ else
...
   end
 end
+
+# We monkey-patch Base.walkdir to use true as default value for follow_symlinks
+# (normally false is the default), in order to "trick" the Documenter code into
+# following those symlinks.
+# See also:
+# https://github.com/JuliaDocs/Documenter.jl/pull/552
+# https://github.com/JuliaLang/julia/blob/master/doc/make.jl#L19
+Base.walkdir(str::String) = Base.walkdir(str; follow_symlinks=true)
```
Comment (fingolfin) [inline docs/documenter_helpers.jl:33]: Hmm, we only need this to make Documenter deal with symlinks *to directories*, right ? I think symlinks to *files* work, don't they?

That would suggest an alternative solution: recursively "closing" directories in the source by creating equally named dirs in the destination (if they are not there already) and then adding symlinks for all the files (and deleting stale symlinks). Quite a bit more work (though not really hard).

The main concern I have with the above is that `Oscar.build_doc` loads this code and then `Base.walkdir` is "tainted" in the whole session.

Of course this is existing code, so none of what I say here is relevant for merging this PR!

Anyway, I'll try and see if perhaps Documenter.jl is willing to just use `follow_symlinks=true` in their code, see https://github.com/JuliaDocs/Documenter.jl/pull/2573

Comment (fingolfin) [reply docs/documenter_helpers.jl:33]: To be clear this is not meant to block this PR

**Review by fingolfin — APPROVED** [review]

**Pull request merged**